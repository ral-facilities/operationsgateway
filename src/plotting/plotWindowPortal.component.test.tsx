import React from 'react';
import { render } from '@testing-library/react';
import PlotWindowPortal from './plotWindowPortal.component';

describe('Plot Window component', () => {
  const TestComponent = () => <div id="test">Test</div>;
  const onClose = jest.fn();
  const mockAddEventListener = jest.fn();
  const mockRemoveEventListener = jest.fn();
  const mockClearTimeout = jest.fn();
  const mockSetTimeout = (func, timeout) => {
    func();
  };
  const mockWindowClose = jest.fn();
  const newDocument =
    global.window.document.implementation.createHTMLDocument();

  Object.defineProperty(window, 'open', {
    value: () => {
      return {
        document: newDocument,
        addEventListener: mockAddEventListener,
        removeEventListener: mockRemoveEventListener,
        close: mockWindowClose,
        clearTimeout: mockClearTimeout,
        setTimeout: mockSetTimeout,
      };
    },
  });

  const createView = () =>
    render(
      <PlotWindowPortal onClose={onClose} title="test title">
        <TestComponent />
      </PlotWindowPortal>
    );

  it('renders child in separate document and initialises event listeners, and handles unmounting correctly', () => {
    const { unmount } = createView();

    expect(newDocument.body).toMatchSnapshot();
    expect(mockAddEventListener).toHaveBeenCalledTimes(2);
    expect(mockAddEventListener).toHaveBeenCalledWith(
      'resize',
      expect.any(Function)
    );
    expect(mockAddEventListener).toHaveBeenCalledWith('beforeunload', onClose);
    expect(newDocument.title).toEqual('OperationsGateway Plot - test title');

    unmount();
    expect(mockRemoveEventListener).toHaveBeenCalledWith(
      'resize',
      expect.any(Function)
    );
    expect(mockWindowClose).toHaveBeenCalled();
  });

  it('changes title on title prop change', () => {
    const { rerender } = createView();

    rerender(
      <PlotWindowPortal onClose={onClose} title="new test title">
        <TestComponent />
      </PlotWindowPortal>
    );

    expect(newDocument.title).toEqual(
      'OperationsGateway Plot - new test title'
    );
  });

  it('removed and re-adds event listeners onClose prop change', () => {
    const { rerender } = createView();

    const newMockOnClose = jest.fn();

    rerender(
      <PlotWindowPortal onClose={newMockOnClose} title="test title">
        <TestComponent />
      </PlotWindowPortal>
    );

    expect(mockRemoveEventListener).toHaveBeenCalledWith(
      'beforeunload',
      onClose
    );
    expect(mockAddEventListener).toHaveBeenCalledWith(
      'beforeunload',
      newMockOnClose
    );
  });

  it('sends a resize event to the main document on popup resize', () => {
    let resizeHandler;
    mockAddEventListener.mockImplementation((event, handler) => {
      if (event === 'resize') {
        resizeHandler = handler;
      }
    });

    const dispatchEventSpy = jest.spyOn(window, 'dispatchEvent');

    createView();

    resizeHandler();

    expect(dispatchEventSpy).toHaveBeenCalledWith(expect.any(Event));
    expect(dispatchEventSpy.mock.calls[0][0].type).toBe(
      'resize OperationsGateway Plot - test title'
    );
  });
});
