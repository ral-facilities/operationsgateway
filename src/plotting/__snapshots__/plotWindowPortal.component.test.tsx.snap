// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plot Window component renders child in separate document and initialises event listeners, and handles unmounting correctly 1`] = `
<body>
  <div>
    <div
      id="test"
    >
      Test
    </div>
  </div>
</body>
`;
